confluent_kafka/__init__.py

```diff
@@ -13,7 +16,633 @@
+                            synonym for the parent configuration entry.
+    :ivar synonyms list: A ConfigEntry list of synonyms and alternate sources for this configuration property.
+
+
+    Source types (.source):
+        :py:const:`CONFIG_SOURCE_UNKNOWN_CONFIG`,
+        :py:const:`CONFIG_SOURCE_DYNAMIC_TOPIC_CONFIG`,
+        :py:const:`CONFIG_SOURCE_DYNAMIC_BROKER_CONFIG`,
+        :py:const:`CONFIG_SOURCE_DYNAMIC_DEFAULT_BROKER_CONFIG`,
+        :py:const:`CONFIG_SOURCE_STATIC_BROKER_CONFIG`,
+        :py:const:`CONFIG_SOURCE_DEFAULT_CONFIG`
+
+    """
+
+    _source_name_by_type = {
+        CONFIG_SOURCE_UNKNOWN_CONFIG: "UNKNOWN_CONFIG",
+        CONFIG_SOURCE_DYNAMIC_TOPIC_CONFIG: "DYNAMIC_TOPIC_CONFIG",
+        CONFIG_SOURCE_DYNAMIC_BROKER_CONFIG: "DYNAMIC_BROKER_CONFIG",
+        CONFIG_SOURCE_DYNAMIC_DEFAULT_BROKER_CONFIG: "DYNAMIC_DEFAULT_BROKER_CONFIG",
+        CONFIG_SOURCE_STATIC_BROKER_CONFIG: "STATIC_BROKER_CONFIG",
+        CONFIG_SOURCE_DEFAULT_CONFIG: "DEFAULT_CONFIG"
+    }
+
+    def __init__(self, name, value,
+                 source=CONFIG_SOURCE_UNKNOWN_CONFIG,
+                 is_read_only=False,
+                 is_default=False,
+                 is_sensitive=False,
+                 is_synonym=False,
+                 synonyms=[]):
+        """
+        This class is typically not user instantiated.
+        """
+        super(ConfigEntry, self).__init__()
+        self.name = name
+        self.value = value
+        self.source = source
+        self.is_read_only = bool(is_read_only)
+        self.is_default = bool(is_default)
+        self.is_sensitive = bool(is_sensitive)
+        self.is_synonym = bool(is_synonym)
+        self.synonyms = synonyms
+
+    def __repr__(self):
+        return "ConfigEntry(%s=\"%s\")" % (self.name, self.value)
+
+    def __str__(self):
+        return "%s=\"%s\"" % (self.name, self.value)
+
+    @classmethod
+    def config_source_to_str(cls, source):
+        """Return string representation of a config source."""
+        return ConfigEntry._source_name_by_type.get(source, "%d?" % source)
+
+
+@functools.total_ordering
+class ConfigResource(object):
+    """
+    Class representing resources that have configs.
+
+
+    :py:const:`RESOURCE_ANY` (used for lookups),
+    :py:const:`RESOURCE_UNKNOWN` (used for reporting),
+    :py:const:`RESOURCE_TOPIC`,
+    :py:const:`RESOURCE_GROUP`,
+    :py:const:`RESOURCE_BROKER`
+
+    Resource names for types:
+      * RESOURCE_TOPIC - topic name
+      * RESOURCE_BROKER - broker id (as str)
+      * RESOURCE_GROUP - group name
+    """
+    _res_name_by_type = {RESOURCE_UNKNOWN: "unknown",
+                         RESOURCE_ANY: "any",
+                         RESOURCE_TOPIC: "topic",
+                         RESOURCE_GROUP: "group",
+                         RESOURCE_BROKER: "broker"}
+    _res_type_by_name = {v: k for k, v in _res_name_by_type.items()}
+
+    def __init__(self, restype, name,
```

Comment:
an enum seems better to me for `restype` too, based on my research just now but no actual prior experience guiding me in what's normal. given it's an input rather than output thing though, not as important as `CONFIG_SOURCE` (but they should be consistent).

Reply:
doin' it